### PR TITLE
[Sema] [CSGen] Solve types through a constraint system

### PR DESCRIPTION
**Pull request by harlanhaskins**
This builds off work done by @CodaFi to implement a constraint solver that will solve for type variables.

### CODE REVIEW

File: Sources/Semantic Analysis/ConstraintSolver.swift [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        system.insert(.equal(returnType1, returnType2), at: 0)
+
+        return solveSystem(system.map(c.withKind))
+      case (.pointer(_), .pointer(_)):
+        // Pointers may unify with any other kind of pointer.
+        return [:]
+      case (.bool, .int):
+        // Boolean values may coerce to integer values (but not vice-versa).
+        return [:]
+      case (_, .any), (.any, _):
```
Comment (CodaFi) [inline Sources/Semantic Analysis/ConstraintSolver.swift]: Add an existential conversion expression and kill this with 🔥

File: Sources/Base/Types.swift [hunk 2]
```diff
@@ -540,7 +540,11 @@ class TypeRefExpr: DeclRefExpr<TypeDecl> {
 
 extension DataType {
   func ref(range: SourceRange? = nil) -> TypeRefExpr {
-    return TypeRefExpr(type: self, name: Identifier(name: "\(self)", range: range), sourceRange: range)
+    let expr = TypeRefExpr(type: self,
+                           name: Identifier(name: "\(self)", range: range),
+                           sourceRange: range)
+    expr.type = self
```
Comment (segiddins) [inline Sources/Base/Types.swift:513]: is this necessary, given we pass the type to the initializer?

Comment (harlanhaskins) [reply Sources/Base/Types.swift:513]: Good catch. I threw this in and forgot to remove it.

File: Sources/Base/ASTContext.swift [hunk 3]
```diff
@@ -281,24 +282,59 @@ public class ASTContext {
...
+    }
+    return true
+  }
   
+  func missingMethodsForConformance(_ decl: TypeDecl, to proto: ProtocolDecl) -> [MethodDecl] {
+    guard let methods = requiredMethods(for: proto) else { return [] }
+    var missing = [MethodDecl]()
+    for method in methods {
+      var impl: MethodDecl?
+      for candidate in decl.methods(named: method.name.name) {
```
Comment (segiddins) [inline Sources/Base/ASTContext.swift]: `impl = decl.methods(named:).first { ahveEqualSigs }`

File: Sources/Base/ASTContext.swift [hunk 4]
```diff
@@ -281,24 +282,59 @@ public class ASTContext {
...
+
   /// - Returns: Whether the expression's type was changed
   @discardableResult
   func propagateContextualType(_ contextualType: DataType, to expr: Expr) -> Bool {
     let canTy = canonicalType(contextualType)
     switch expr.semanticsProvidingExpr {
     case let expr as NumExpr:
       if case .int = canTy {
-        expr.type = contextualType
+        expr.type = canTy
```
Comment (segiddins) [inline Sources/Base/ASTContext.swift]: didn't we want to continue propogating the sugared type?

Comment (harlanhaskins) [reply Sources/Base/ASTContext.swift]: Oops.

File: Sources/Base/Functions.swift [hunk 5]
```diff
@@ -65,10 +65,20 @@ class FuncDecl: Decl { // func <id>(<id>: <type-id>) -> <type-id> { <expr>* }
     self.returnType = returnType
     self.hasVarArgs = hasVarArgs
     self.isPlaceholder = isPlaceholder
-    super.init(type: .function(args: args.map { $0.type }, returnType: returnType.type!),
+    var allValid = true
+    for arg in args {
```
Comment (segiddins) [inline Sources/Base/Functions.swift]: `!args.contains {}`

File: Sources/Base/ASTContext.swift (hunk 3 above)
Comment (harlanhaskins) [reply Sources/Base/ASTContext.swift]: 😄

File: Sources/Base/Functions.swift (hunk 5 above)
Comment (harlanhaskins) [reply Sources/Base/Functions.swift]: 😄

File: Sources/Semantic Analysis/ConstraintGenerator.swift [hunk 6]
```diff
@@ -257,11 +257,12 @@ final class ConstraintGenerator: ASTTransformer {
     visit(expr.lhs)
     let lhsGoal = self.goal
 
-    system.constrainEqual(expr.decl!, lhsGoal)
-    let tau = env.freshTypeVariable()
+    guard case .tuple(let fields) = lhsGoal else {
+      return
```
Comment (CodaFi) [inline Sources/Semantic Analysis/ConstraintGenerator.swift]: `fatalError`?

File: Sources/Base/ASTContext.swift (hunk 4 above)
Comment (CodaFi) [reply Sources/Base/ASTContext.swift]: Ideally we propagate nothing and this entire method goes the way of the dinosaurs.

Comment (harlanhaskins) [reply Sources/Base/ASTContext.swift]: Every other one of them propagates the `canTy`...

Comment (segiddins) [reply Sources/Base/ASTContext.swift]: only as of this PR...

File: Sources/Semantic Analysis/Sema.swift [hunk 7]
```diff
@@ -890,7 +879,7 @@ class Sema: ASTTransformer, Pass {
         return
       }
     }
-
+
```
Comment (segiddins) [inline Sources/Semantic Analysis/Sema.swift:882]: whitespace

File: trill.xcodeproj/project.xcworkspace/xcshareddata/trill.xcscmblueprint [hunk 8]
```diff
@@ -24,7 +24,7 @@
       "DVTSourceControlWorkspaceBlueprintRemoteRepositoryIdentifierKey" : "32E9A4F24A891D682A50DDAB6DE720492519A2F6"
     },
     {
-      "DVTSourceControlWorkspaceBlueprintRemoteRepositoryURLKey" : "github.com:trill-lang\/trill.git",
+      "DVTSourceControlWorkspaceBlueprintRemoteRepositoryURLKey" : "github.com:CodaFi\/trill.git",
```
Comment (segiddins) [inline trill.xcodeproj/project.xcworkspace/xcshareddata/trill.xcscmblueprint]: uh no

File: Sources/Base/Types.swift (hunk 2 above)
Comment (segiddins) [reply Sources/Base/Types.swift:513]: this never got removed...

**Pull request merged**